pkgs/development/python-modules/qiskit-terra/default.nix

```diff
@@ -0,0 +1,109 @@
+{ lib
+, pythonOlder
+, buildPythonPackage
+, fetchFromGitHub
+, cython
+, python
+, dill
+, jsonschema
+, numpy
+, marshmallow
+, marshmallow-polyfield
+, matplotlib
+, networkx
+, ply
+, psutil
+, scipy
+, sympy
+  # test requirements
+, ddt
+, hypothesis
+, ipywidgets
+, nbformat
+, nbconvert
+, pillow
+, pydot
+, pygraphviz
+, pylatexenc
+, pytestCheckHook
+}:
+
+buildPythonPackage rec {
+  pname = "qiskit-terra";
+  version = "0.12.0";
+
+  disabled = pythonOlder "3.5";
+
+  src = fetchFromGitHub {
+    owner = "Qiskit";
+    repo = pname;
+    rev = version;
+    sha256 = "1yarfziy2w8n1d7zyyxykfs68608j8md4kwfyhbyc6wy483fk9sy";
+  };
+
+  nativeBuildInputs = [ cython ];
+
+  propagatedBuildInputs = [
+    dill
+    jsonschema
+    numpy
+    marshmallow
+    marshmallow-polyfield
+    matplotlib
+    networkx
+    ply
+    psutil
+    scipy
+    sympy
+  ];
+
+  checkInputs = [
+    ddt
+    hypothesis
+    ipywidgets
+    nbformat
+    nbconvert
+    pillow
+    pydot
+    pygraphviz
+    pylatexenc
+    pytestCheckHook
+  ];
+  pythonImportsCheck = [
+    "qiskit"
+    "qiskit.transpiler.passes.routing.cython.stochastic_swap.swap_trial"
+  ];
+  dontUseSetuptoolsCheck = true;
+  disabledTests = [
+    "test_long_name"  # images differ for some reason
+    "test_jupyter_jobs_pbars" # needs IBMQ provider
+  ];
+  pytestFlagsArray = [
+    "--ignore=test/randomized/test_transpiler_equivalence.py" # collection requires qiskit-aer, which would cause circular dependency
+  ];
+  preCheck = ''
+    # Tests located in $PACKAGEDIR/test
+    export PACKAGEDIR=$out/${python.sitePackages}
+    echo "Moving Qiskit test files to package directory"
+    # cp $TMP/source/.stestr.conf $PACKAGEDIR
+    cp -r $TMP/source/test $PACKAGEDIR
+    cp -r $TMP/source/examples $PACKAGEDIR
+    cp -r $TMP/source/qiskit/schemas/examples $PACKAGEDIR/qiskit/schemas/
+
+    # run tests from Nix's $out path
+    pushd $PACKAGEDIR
+  '';
+  postCheck = ''
+    popd
+  '';
```

Comment:
it may be unnecessary to copy all the stuff to $out after https://github.com/NixOS/nixpkgs/pull/80595

Reply:
With preCheck/postCheck commented out, pytest is run from the wrong directory so it fails.

I move the files to $PACKAGEDIR because pytest doesn't find the cythonized modules correctly otherwise. And there are some test files that expect to be located IN the test directories.

Reply:
#80595 just relates to importcheck, whereas this is for running ``pytest``

Reply:
yea, but the import check changes the directory to $out, which is why you no longer had the tests that were located in source.

Reply:
Didn't realize #80595 was targeting ``staging`` vs ``master``, so I assumed I had it. I'm trying it now w/ that commit cherry-picked.

Reply:
Tried not copying to $out (cherry-picking #80595), and it failed because it was missing the cythonized modules. I tried preCheck="cd test" too and that didn't work. I still don't have a good handle on why Nix can't find the cythonized modules in the CheckPhase.

Reply:
But this unfortunately somewhat-convoluted system at least builds. I'm open to suggestions, but it seems this works for now.

Reply:
c extensions makes sense

Reply:
unfortunately the pythonImportsCheck change affects almost the whole python stack, so all of python packages essentially needs to be rebuilt